### PR TITLE
🐛 Fix node_reuse and remediation flakiness

### PR DESCRIPTION
**Pull request by mboukhalfa**
Fix node_reuse flakiness such as:
https://jenkins.nordix.org/view/Metal3/job/airship_metal3io_capi_m3_master_e2e_test_centos/24/console
https://jenkins.nordix.org/view/Metal3/job/airship_metal3io_capi_m3_master_e2e_test_centos/23/console
https://jenkins.nordix.org/view/Metal3/job/airship_metal3io_capi_m3_master_e2e_test_centos/18/console
https://jenkins.nordix.org/view/Metal3/job/airship_metal3io_capi_m3_master_e2e_test_ubuntu/30/console
https://jenkins.nordix.org/view/Metal3/job/airship_metal3io_capi_m3_master_e2e_test_ubuntu/16/console

### CODE REVIEW

**Review by namnx228 — COMMENTED** [review]
LGTM

**Review by namnx228 — COMMENTED** [review]
/lgtm

File: test/e2e/node_reuse_test.go [hunk 1]
```diff
@@ -455,15 +513,20 @@ func updateNodeReuse(nodeReuse bool, m3machineTemplateName string, clusterClient
 	Expect(m3machineTemplate.Spec.NodeReuse).To(BeTrue())
 }
 
-func untaintNodes(clientSet *kubernetes.Clientset, nodes *corev1.NodeList, taint *corev1.Taint) {
+func untaintNodes(clientSet *kubernetes.Clientset, nodes *corev1.NodeList, taint *corev1.Taint) (count int) {
+	count = 0
 	for i := range nodes.Items {
+		Logf("Untainting node %v ...", nodes.Items[i].Name)
```
Comment (kashifest) [inline test/e2e/node_reuse_test.go:519]: do we need this loop to untaint? Isn't it so that we can untaint all nodes at once using node labels?

Comment (mboukhalfa) [reply test/e2e/node_reuse_test.go:519]: We can either wait by adding sleep for some time or loop until untainting the node. for some reason the taint is not present just after provisioning

File: test/e2e/remediation_test.go [hunk 2]
```diff
@@ -115,8 +120,15 @@ func test_remediation() {
...
-		bmhs := getAllBmhs(ctx, bootstrapClient, namespace, specName)
-		g.Expect(filterBmhsByProvisioningState(bmhs, bmh.StateProvisioned)).To(HaveLen(2))
+		bmhs, err := getAllBmhs(ctx, bootstrapClient, namespace, specName)
+		if err != nil {
+			Logf("Error: %v", err)
+			return err
+		}
+		provisioningBMHs := filterBmhsByProvisioningState(bmhs, bmh.StateProvisioned)
+		if len(provisioningBMHs) != 2 {
+			return errors.New("Waiting for 2 BMHs to be Provisioned")
```
[the remark this answers is not in the document]
Comment (mboukhalfa) [reply test/e2e/remediation_test.go:130]: Eventually here expect an error return

**Review by Rozzii — APPROVED** [review]

File: test/e2e/node_reuse_test.go [hunk 3]
```diff
@@ -193,17 +242,25 @@ func node_reuse() {
...
-	err = targetClusterClient.Patch(ctx, &ctrlplane, client.RawPatch(types.MergePatchType, patch))
-	Expect(err).To(BeNil(), "Failed to set up KCP maxSurge to 1")
+	Eventually(
+		func() error {
+			return targetClusterClient.Patch(ctx, &ctrlplane, client.RawPatch(types.MergePatchType, patch))
+		}, e2eConfig.GetIntervals(specName, "wait-patch")...,
+	).Should(Succeed(), "Failed to set up KCP maxSurge to 1")
+
+	By("Untaint all CP nodes")
+	// The rest of CP nodes may take time to be untaintable
```
Comment (smoshiur1237) [inline test/e2e/node_reuse_test.go:252]: I am not convinced with this step. In previous step you assure that env has 3 updated CP bmh and machines are in running state.  If CPs is running, it doesn't need untaint loop at all.

File: test/e2e/node_reuse_test.go (hunk 1 above)
Comment (smoshiur1237) [reply test/e2e/node_reuse_test.go:519]: I am not convinced with this step. In previous step you assure that env has 3 updated CP bmh and machines are in running state. If CPs is running, it doesn't need untaint loop at all.

Comment (mboukhalfa) [reply test/e2e/node_reuse_test.go:519]: @kashifest you mean `kubectl taint nodes --all node-role.kubernetes.io/master-` not sure if that feasible but I think at the end it end updating node by node.

@smoshiur1237 can you please refer to the step ? if you refer to the last cp untaint although all cp are running we we try to scale down the kcp to 1 if not all the nodes are untainted there are a probability that the nodes that will try still have the taint and the other nodes that are deprovisioning cannot migrate the pods to it

Comment (kashifest) [reply test/e2e/node_reuse_test.go:519]: > @kashifest you mean `kubectl taint nodes --all node-role.kubernetes.io/master-` not sure if that feasible but I think at the end it end updating node by node.
Yes and IMO thats what the command is suitable for, feasibility.

Comment (smoshiur1237) [reply test/e2e/node_reuse_test.go:519]: I meant for the last untaint step.

Comment (mboukhalfa) [reply test/e2e/node_reuse_test.go:519]: @smoshiur1237 Yes, if did not add that the last cp node that was upgraded will still have the taint. and the next step is scalling down kcp to 1 if that one was the node that has the untaint scaling down will hangs

**Review by namnx228 — APPROVED** [review]
LGTM

**Review by Rozzii — APPROVED** [review]

File: test/e2e/node_reuse_test.go [hunk 4]
```diff
@@ -103,22 +111,35 @@ func node_reuse() {
 		}, e2eConfig.GetIntervals(specName, "wait-machine-deleting")...,
 	).Should(Succeed())
 
-	Byf("Wait until %d BMH is in deprovisioning state", numberOfWorkerBmh)
+	Byf("Wait until 1 BMH is in deprovisioning state")
 	deprovisioning_bmh := []bmo.BareMetalHost{}
 	Eventually(
-		func() int {
+		func() error {
```
Comment (kashifest) [inline test/e2e/node_reuse_test.go:117]: why error instead of init?

Comment (mboukhalfa) [reply test/e2e/node_reuse_test.go:117]: so we can keep retry if any flake rise

**Review by kashifest — APPROVED** [review]

**Review by kashifest — APPROVED** [review]

**Review by kashifest — APPROVED** [review]

**Pull request merged**